shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java

```diff
@@ -0,0 +1,60 @@
+package io.flutter.embedding.engine.renderer;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.annotation.Config;
+
+import io.flutter.embedding.engine.FlutterJNI;
+import io.flutter.embedding.engine.renderer.FlutterRenderer;
+import io.flutter.embedding.engine.renderer.OnFirstFrameRenderedListener;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+@Config(manifest=Config.NONE)
+@RunWith(RobolectricTestRunner.class)
+public class FlutterRendererTest {
+  FlutterJNITest jniUnderTest;
+  FlutterRenderer rendererUnderTest;
+
+  @Before
+  public void setUp() {
+    jniUnderTest = new FlutterJNITest();
+    rendererUnderTest = new FlutterRenderer(jniUnderTest);
+  }
+
+  @Test
+  public void firstFrameListenerCanRemoveThemselves() {
+    AtomicInteger callbackCalled = new AtomicInteger(0);
+    OnFirstFrameRenderedListener callback = new OnFirstFrameRenderedListener() {
+      @Override
+      public void onFirstFrameRendered() {
+        callbackCalled.incrementAndGet();
+        rendererUnderTest.removeOnFirstFrameRenderedListener(this);
+      };
+    };
+
+    rendererUnderTest.addOnFirstFrameRenderedListener(callback);
+    jniUnderTest.onFirstFrame();
```

Comment:
I would recommend that in all tests we clearly demarcate between setup/execution/verification. In this case everything up to and including the `addOnFirstFrameRenderedListener` is part of the setup, the `onFirstFrame()` invocation is the execution, and everything that follows is verification.

In addition to placing a blank line and maybe a comment between these 2 lines above, you could also slightly refactor what comes below. I dont think you actually need to interleave 2 `onFirstFrame()`s and 2 `assert`s. You can simply invoke `onFirstFrame()` twice in a row, and then verify that the callback was invoked exactly 1 time.

Reply:
I think it's still worth checking for the actual timing (e.g. that we don't have false positives on onFirstFrame only calling listeners on second invocation or something)